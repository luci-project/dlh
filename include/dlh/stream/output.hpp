#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
