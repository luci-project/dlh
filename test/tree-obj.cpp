#include <dlh/stream/output.hpp>
#include <dlh/container/tree.hpp>

#include "foo.hpp"

int main(int argc, const char *argv[]) {
	(void) argc;
	(void) argv;

	cout << "TreeSet:";
	TreeSet<Foo, FooComp> s = { 888, 999 };
	s.insert(13);
	s.insert(Foo(3));
	Foo fourtytwo(42);
	s.insert(fourtytwo);
	s.insert(23);
	s.emplace(7);

	s.emplace(Foo(13));
	s.insert(1549);
	const Foo numberofthebeast(666);
	s.emplace(numberofthebeast);
	s.insert(3085);
	s.insert(204);
	s.erase(numberofthebeast);

	s.erase(3085);
	s.insert(1572877);
	auto it1 = s.find(204);
	if (it1) {
		auto && tmp = s.extract(it1);
		tmp.value().id = 205;
		s.insert(move(tmp));
	}
	s.insert(13);
	s.resize(128);
	s.insert(32);
	auto it2 = s.find(7);
	if (it2) {
		auto && tmp = s.extract(it2);
		tmp.value().id = 13;
		s.insert(move(tmp));
	}
	s.insert(42);
	s.insert(52);

	#ifndef NDEBUG
	s.check();
	#endif

	TreeSet<Foo, FooComp> t(s);
	for (auto i : t)
		cout << ' ' << i.id;
	cout << endl;

	int n = 42;
	cout << " - Lowest: " << s.lowest()->id << endl;
	cout << " - Lower than " << n << ": " << s.lower(n)->id << endl;
	cout << " - Lower than or equal to " << n << ": " << s.floor(n)->id << endl;
	cout << " - Higher than or equal to " << n << ": " << s.ceil(n)->id << endl;
	cout << " - Higher than " << n << ": " << s.higher(n)->id << endl;
	cout << " - Highest: " << s.highest()->id << endl;
	cout << endl;

	return 0;
}
