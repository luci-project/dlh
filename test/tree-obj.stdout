TreeSet: 3 13 23 32 42 52 205 888 999 1549 1572877
 - Lowest: 3
 - Lower than 42: 32
 - Lower than or equal to 42: 42
 - Higher than or equal to 42: 42
 - Higher than 42: 52
 - Highest: 1572877
